podcastfy/core/podcast.py

```diff
@@ -0,0 +1,410 @@
+        """Save the transcript to a file."""
+        with open(filepath, 'w') as f:
+            f.write(str(self))
+
+    def __str__(self) -> str:
+        """Convert the transcript to a string representation."""
+        lines = []
+        for segment in self.segments:
+            lines.append(f"{segment.speaker}: {segment.text}")
+
+        metadata_str = "\n".join([f"{key}: {value}" for key, value in self.metadata.items()])
+
+        return f"Metadata:\n{metadata_str}\n\nTranscript:\n" + "\n".join(lines)
+
+
+class AudioSegment:
+    """Represents an audio segment of the podcast."""
+
+    def __init__(self, filepath: Path, length_ms: int, transcript_segment: Optional[TranscriptSegment] = None):
+        self.filepath = filepath
+        self.length_ms = length_ms
+        self.transcript_segment = transcript_segment
+        self._audio: Optional[PydubAudioSegment] = None
+
+    @property
+    def audio(self) -> PydubAudioSegment:
+        """Lazy-load the audio segment."""
+        if self._audio is None:
+            self._audio = PydubAudioSegment.from_file(self.filepath)
+            if len(self._audio) != self.length_ms:
+                raise ValueError(
+                    f"Audio file length ({len(self._audio)}ms) does not match specified length ({self.length_ms}ms)")
+        return self._audio
+
+
+def podcast_stage(func):
+    """Decorator to manage podcast stage transitions."""
+
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        current_method = self._next_stage_methods[self.state]
+        if current_method != func and not self._reworking:
+            print(f"Cannot execute {func.__name__} in current state {self.state.name}. Skipping.")
+            return
+
+        try:
+            result = func(self, *args, **kwargs)
+            next_state = next((state for state, method in self._next_stage_methods.items() if method == func), None)
+            self.state = next_state if next_state else self.state
+            return result
+        except Exception as e:
+            print(f"Error in {func.__name__}: {str(e)}")
+            raise
+
+    return wrapper
+
+
+class Podcast:
+    """Main class for podcast creation and management."""
+
+    def __init__(self, content: str, llm_backend: LLMBackend, tts_backend: TTSBackend, default_tts_n_jobs: int = 1):
+        """
+        Initialize a new Podcast instance.
+
+        Args:
+            content (str): The raw content to be processed into a podcast.
+            llm_backend (LLMBackend): The language model backend for generating the transcript.
+            tts_backend (TTSBackend): The text-to-speech backend for converting text to audio.
+            default_tts_n_jobs (int, optional): The default number of concurrent jobs for TTS processing.
+                Defaults to 1.
+
+        Raises:
+            logging.warning: If using a synchronous TTS backend with multiple jobs.
+        """
+        self.content = content
```

Comment:
podcastfy v0.2.1+ is now multimodal, meaning content should be a list of elements, e.g. text or image for now but video and audio too in the future.

Reply:
Indeed!